pkg/ebda/ebda.go

```diff
@@ -0,0 +1,120 @@
+		return 0, fmt.Errorf("error seeking to memory offset 0x%X for the EBDA pointer, got: %v", EBDAAddressOffset, err)
+	}
+
+	// read the ebda pointer
+	if err := binary.Read(f, binary.LittleEndian, &ep); err != nil {
+		return 0, fmt.Errorf("unable to read EBDA Pointer: %v", err)
+	}
+
+	if ep == 0 {
+		return 0, errors.New("ebda offset is 0! unable to proceed")
+	}
+	// The EBDA pointer tends to be shifted 4 bits down.
+	// For reference: https://wiki.osdev.org/Memory_Map_(x86)
+	return int64(ep) << 4, nil
+}
+
+// ReadEBDA assumes it's been passed in /dev/mem and searches for and reads the EBDA.
+func ReadEBDA(f io.ReadSeeker) (*EBDA, error) {
+	var err error
+	var eSize int8
+	e := &EBDA{}
+
+	if e.BaseOffset, err = findEBDAOffset(f); err != nil {
+		return nil, err
+	}
+
+	if _, err = f.Seek(e.BaseOffset, io.SeekStart); err != nil {
+		return nil, fmt.Errorf("error seeking to memory offset 0x%X for the start of the EBDA, got: %v", e.BaseOffset, err)
+	}
+
+	// Read length
+	if err := binary.Read(f, binary.LittleEndian, &eSize); err != nil {
+		return nil, fmt.Errorf("error reading EBDA length, got: %v", err)
+	}
+	e.Length = int64(eSize) << 10
+
+	// Find the size of this segment if the first byte is not set.
+	// We assume that the EBDA never crosses a segment for now.
+	// This is a reasonable assumption in most cases, since most EBDAs I've seen are in the 0x9XXXX range,
+	// and the 0xA0000 segment is off limits.
+	if e.Length == 0 {
+		e.Length = 0x10000 - (e.BaseOffset & 0xffff)
+	}
+	e.Buf = make([]byte, e.Length)
+
+	if _, err = f.Seek(e.BaseOffset, io.SeekStart); err != nil {
+		return nil, fmt.Errorf("error seeking to memory offset 0x%X for the start of the EBDA, got: %v", e.BaseOffset, err)
+	}
+	if err = binary.Read(f, binary.LittleEndian, e.Buf); err != nil {
+		return nil, fmt.Errorf("error reading EBDA region, tried to read from 0x%X of size 0x%X, got %v", e.BaseOffset, e.Length, err)
+	}
+
+	return e, nil
+}
+
+// WriteEBDA assumes it's been passed in /dev/mem and searches for and writes to the EBDA.
+func WriteEBDA(e *EBDA, f io.ReadWriteSeeker) error {
+	var err error
+	if e.BaseOffset == 0 {
+		// Don't know why this is 0, but try to find one from /dev/mem
+		if e.BaseOffset, err = findEBDAOffset(f); err != nil {
+			return err
+		}
+	}
+
+	if e.Length&0x3ff != 0 {
+		// Length can't be encoded as KiB.
+		return fmt.Errorf("length is not an integer multiple of 1 KiB, got 0x%X", e.Length)
+	}
+	if e.Length != int64(len(e.Buf)) {
+		return fmt.Errorf("length field is not equal to buffer length. length field: 0x%X, buffer length: 0x%X", e.Length, len(e.Buf))
+	}
+
+	// Update length field in buffer
+	eLen := e.Length >> 10
+	if eLen > 0xFF {
+		// Length is too big to be written into one byte. fail.
+		return fmt.Errorf("length is greater than 255 KiB, cannot be marshalled into one byte. Length: %dKiB", eLen)
+	}
+
+	// Cheat since it's only one byte
+	e.Buf[0] = byte(eLen)
```

Comment:
If you increase the size of the EBDA, would you risk over writing stuff after it? Likewise, if you decrease the size of the EBDA, don't you risk chopping stuff off? Can you just leave this byte as it was originally?

Reply:
Can you just get rid of it altogether to prevent people from shooting themselves in the foot?

Reply:
No I want to leave the possibility for extending the ebda if there's nothing behind it.